### ISSUE TITLE
wallet: Make rpc result structure independent of return value

### ISSUE THREAD

**Issue by MarcoFalke** [bitcoin/bitcoin#17155]
I think that the RPC result structure (i.e. the potentially nested JSON types) should not depend on the return value, as this makes the deserialization from json to a language-specific struct harder.

For example, `gettransaction` has a key `"generated"`, that can never be false. It either exists (and is true) or does not exists (and is implicitly false).

`generated` (and other fields in `gettransaction`) should be changed to always return the value.

**Comment by MarcoFalke** [comment]
The code is here

https://github.com/bitcoin/bitcoin/blob/e180be49d7d802d05b1850924be97fe27d353ffe/src/wallet/rpcwallet.cpp#L133-L134

**Comment by theStack** [comment]
Other fields in `gettransaction` not always present are e.g. `blockhash`, `blockindex` and `blocktime` -- they only exist for confirmed transactions, for unconfirmed the `trusted` field is generated instead (snippet starts one line below yours about `generated`, by the way):

https://github.com/bitcoin/bitcoin/blob/e180be49d7d802d05b1850924be97fe27d353ffe/src/wallet/rpcwallet.cpp#L135-L145

So it doesn't seem too trivial to solve this (what values should `blockhash`, `blockindex` and `blocktime` get for unconfirmed transactions?). Or is it enough for a first step to approach this just for boolean values?

**Comment by MarcoFalke** [comment]
> what values should blockhash, blockindex and blocktime get for unconfirmed transactions?

I think it is fine to solve it by mentioning it in the help text, which is missing right now.

**Comment by MarcoFalke** [comment]
Similarly, `invovesWatchonly` should probably not be included for non-watchonly wallets, but always be included for watchonly wallets. See also slightly related #16383 which changed the fallbacks for RPC inputs when the wallet is watchonly.

**Comment by theStack** [comment]
Okay it seems I got something wrong here -- is it not the goal to have a fixed result structure, i.e. all fields are always present (basically meaning that there are no conditional `.pushKV()` calls anymore), to get rid of all "only present if...." fields?

**Comment by MarcoFalke** [comment]
You can set `blockhash` to a JSON-None value to keep the structure static. This will make the RPC response slightly larger, but is preferable I think.

**Comment by theStack** [comment]
> You can set `blockhash` to a JSON-None value to keep the structure static. This will make the RPC response slightly larger, but is preferable I think.

That would make sense. That "JSON-None value" can be created with the empty constructur of UniValue `UniValue()`, leading to `null`, as I just found out (as information for anyone wanting to work on this, I'm not sure if I will have time before end of the month).

What confused me was this part:
> Similarly, `invovesWatchonly` should probably not be included for non-watchonly wallets, but always be included for watchonly wallets. 

Shouldn't this field `involvesWatchonly` also always be generated? Right now it is only add to the response if it is true.

**Comment by practicalswift** [comment]
Concept ACK: making the structure independent of return value makes the life easier for our users and the cost of doing so is essentially zero

**Comment by nijynot** [comment]
I can take a look at this, seems like a good first issue.

**Comment by laanwj** [comment]
Concept ACK *in general*. I can think of only one place where it makes a considerable difference in structure size, and was intentional: for the message type statistics in `getpeerinfo`.

**Comment by fjahr** [comment]
Concept ACK

I fully agree that static response structures make the lives of people developing against the RPC easier. While responses may be heavier we can overall probably reduce the number of LoC and making this a general rule prevents future discussions whether certain RPC return values could be made conditional, which mostly seems like a waste of energy IMO.

**Comment by ryanofsky** [comment]
> I fully agree that static response structures make the lives of people developing against the RPC easier.

I think the key advantage of static structures is reduced ambiguity. If field is set conditionally and absent, it's ambiguous whether the field has no value, or the backend just doesn't support returning it. It can make it harder than it needs to be write an RPC client that isn't tied to specific bitcoind version.

**Comment by MarcoFalke** [comment]
> clients are expecting missing keys and we should avoid breaking changes. So we will end up with a non-consistent API anyway. For this reason I also don't think this is a good first issue.

Good point @promag . Removed "good first issue" for now.

**Comment by MarcoFalke** [comment]
Also good point by @ryanofsky . A living example for that is https://github.com/bitcoin/bitcoin/pull/15077#issuecomment-452077683

**Comment by sipa** [comment]
My view is that well-documented cases where values are conditionally present based on other fields (e.g. getaddressinfo only contains the sub-structure that recurses into p2sh when it also indicates you have a p2sh address). That avoids the issue @ryanofsky brings up above where you don't know if something is not present because the software doesn't support it, or because it is intended to be missing.

In other words: I think it would be ideal that fields being missing or not itself should never be used to convey information.

**Comment by promag** [comment]
@MarcoFalke and what would be `decoded` value in `gettransaction verbose=false` result?
